receiver/filereceiver/file_reader.go

```diff
@@ -58,25 +104,193 @@ func (fr fileReader) readAll(ctx context.Context) error {
+	if err != nil {
+		return fmt.Errorf("failed to read line from input file: %w", err)
+	}
+	traces, err := fr.unmarshaler.tracesUnm.UnmarshalTraces([]byte(line))
+	if err != nil {
+		return fmt.Errorf("failed to unmarshal traces: %w", err)
+	}
+	err = fr.timer.wait(ctx, getFirstTimestampFromTraces(traces))
+	if err != nil {
+		return fmt.Errorf("readLine interrupted while waiting for timer: %w", err)
+	}
+	return fr.consumer.tracesConsumer.ConsumeTraces(ctx, traces)
+}
+
+// readMetricLine reads the next line in the file, converting it into metrics and
 // passing it to the the consumer member.
-func (fr fileReader) readLine(ctx context.Context) error {
+func (fr fileReader) readMetricLine(ctx context.Context) error {
 	line, err := fr.stringReader.ReadString('\n')
 	if err != nil {
 		return fmt.Errorf("failed to read line from input file: %w", err)
 	}
-	metrics, err := fr.unm.UnmarshalMetrics([]byte(line))
+	metrics, err := fr.unmarshaler.metricsUnm.UnmarshalMetrics([]byte(line))
+	if err != nil {
+		return fmt.Errorf("failed to unmarshal metrics: %w", err)
+	}
+	err = fr.timer.wait(ctx, getFirstTimestampFromMetrics(metrics))
+	if err != nil {
+		return fmt.Errorf("readLine interrupted while waiting for timer: %w", err)
+	}
+	return fr.consumer.metricsConsumer.ConsumeMetrics(ctx, metrics)
+}
+
+// readAllChunks reads the next chunk of data where each chunk is prefixed with
+// the size of the data chunk.
+func (fr fileReader) readAllChunks(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+			var err error
+			switch {
+			case fr.consumer.tracesConsumer != nil:
+				err = fr.readTraceChunk(ctx)
+			case fr.consumer.metricsConsumer != nil:
+				err = fr.readMetricChunk(ctx)
+			case fr.consumer.logsConsumer != nil:
+				err = fr.readLogChunk(ctx)
+			}
+
+			if err != nil {
+				if errors.Is(err, io.EOF) {
+					return nil
+				}
+				return err
+			}
+		}
+	}
+}
+
+func (fr fileReader) readMetricChunk(ctx context.Context) error {
+	var sz int32
+	err := binary.Read(fr.stringReader, binary.BigEndian, &sz)
+	if err != nil {
+		return err
+	}
+	dataBuffer := make([]byte, sz)
+	err = binary.Read(fr.stringReader, binary.BigEndian, &dataBuffer)
+	if err != nil {
+		return fmt.Errorf("failed to read line from input file: %w", err)
+	}
+	metrics, err := fr.unmarshaler.metricsUnm.UnmarshalMetrics(dataBuffer)
 	if err != nil {
 		return fmt.Errorf("failed to unmarshal metrics: %w", err)
 	}
-	err = fr.timer.wait(ctx, getFirstTimestamp(metrics))
+	err = fr.timer.wait(ctx, getFirstTimestampFromMetrics(metrics))
+	if err != nil {
+		return fmt.Errorf("readLine interrupted while waiting for timer: %w", err)
+	}
+	return fr.consumer.metricsConsumer.ConsumeMetrics(ctx, metrics)
+}
+
+func (fr fileReader) readTraceChunk(ctx context.Context) error {
+	var sz int32
+	err := binary.Read(fr.stringReader, binary.BigEndian, &sz)
+	if err != nil {
+		return err
+	}
+
+	dataBuffer := make([]byte, sz)
```

Comment:
This is to accurately decode the method of writing the fileexporter uses for proto format. The filexporter currently already supports 2 distinct writing strategies. (1) Writing line by line or (2) writing chunks that are prepended with the size of the chunk (for proto format)

This added functionality now supports the second type of file writing in the exporter.  We can't use EOL here because in this case the writer does not include any newlines so decoding would not be accurate

Reply:
@atoulme Yes, the old existing code was weird!